### PR TITLE
Implemented Level 1 Of Rag

### PR DESCRIPTION
**Pull request by ivanleomk**
Implemented Level 1 of Rag using lancedb and openai-embedding model. 

----

| <a href="https://ellipsis.dev" target="_blank">Ellipsis</a> | :rocket: This PR description was created by [Ellipsis](https://www.ellipsis.dev) for commit fbcc068f32cd33ce9bbf9688f1017a702403b5d0.  | 
|--------|--------|

### Summary:
This PR introduces Level 1 of Rag implementation using lancedb and OpenAI's text-embedding model, adding a new text document and scripts for data embedding.

**Key points**:
- Added a new markdown file `superlinear-returns.md` containing a text document.
- Created `embed.py` to handle the embedding of the text data using OpenAI's text-embedding model.
- Defined two new Pydantic models, `Document` and `TextChunk`, in `helpers/models.py`.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

**Comment by ivanleomk** [comment]
There's a bug now whereby if you use the `TextChunk` model and return it in the `chunk_text`, when we insert it into the database, it'll just create a vector of 0s

```
>>> db = lancedb.connect("./db")
>>> tbl = db.open_table("paul_graham")
>>> tbl.to_pandas()
    id  doc_id  chunk_num  start_pos  end_pos                                               text                                             vector
0    0       0          0          0     1024  One of the most important things I didn't unde...  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
1    1       0          1       1024     2048  initely should, because this will be the wave ...  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
2    2       0          2       2048     3072  early on, when the absolute numbers are still ...  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
3    3       0          3       3072     4096  stoms to deal with, but we don't have many cus...  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
4    4       0          4       4096     5120  eople were startlingly more knowledgeable than...  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
..  ..     ...        ...        ...      ...                                                ...                                                ...
73  73       1         53      54272    55296  t. Seek out the people who increase your energ...  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
74  74       1         54      55296    56320  me bitter.\n\nIt's ok to want to impress other...  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
75  75       1         55      56320    57344  d be on "curiosity."\n\nThat doesn't translate...  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
76  76       1         56      57344    58368  own to ability and interest. Can you find a ki...  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
77  77       1         57      58368    59392   want to do great work, or not? Now you have t...  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...
```

Not sure how to fix this but yielding a dict seems to work nicely for now so keeping it as a temporary fix.

File: helpers/db.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+import lancedb
+from helpers.models import TextChunk, Document
+from typing import Iterable, Literal
+from pathlib import Path
+
+
+def create_db(path):
```
Comment (jxnl) [inline helpers/db.py:7]: we dont need functions that just call the a function 

def f(a):
   return g(a)

File: helpers/db.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import lancedb
```
Comment (jxnl) [inline helpers/db.py]: this file is not db.py. cause it responsible for adding data. 
lets keep everything in a single file

Comment (ivanleomk) [reply helpers/db.py]: Sure thing, will make the changes to put everything in one single file so it's simpler.

File: rag_app/cli.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import typer
+import rag_app.query as QueryApp
+
+app = typer.Typer(
+    name="instructor-ft",
```
Comment (jxnl) [inline rag_app/cli.py]: typo

File: rag_app/cli.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+    help="A CLI for fine-tuning OpenAI's models",
+)
+
+app.add_typer(
+    QueryApp.app, name="db", help="Commands to help interact with your local lancedb"
```
Comment (jxnl) [inline rag_app/cli.py]: commands should be verbs

query >> "db"

File: rag_app/query.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        .data[0]
+        .embedding
+    )
+
+    results: List[TextChunk] = (
+        table.search(query_vector).limit(2).to_pydantic(TextChunk)
+    )
+
+    print("=========================")
+    for idx, result in enumerate(results):
```
Comment (jxnl) [inline rag_app/query.py]: use rich.Table

**Comment by jxnl** [comment]
please make stacked PRS rather than pushing into one branch

**Comment by ivanleomk** [comment]
Removed the rag-app cli files and will make a stacked PR for that

**Pull request merged**